lib/srv/discovery/discovery.go

```diff
@@ -0,0 +1,174 @@
+	// ec2Installer is used to start the installation process on discovered EC2 nodes
+	ec2Installer *server.SSMInstaller
+}
+
+func New(ctx context.Context, cfg *Config) (*Server, error) {
+	s := &Server{
+		Config: cfg,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component:       teleport.ComponentDiscovery,
+			trace.ComponentFields: logrus.Fields{},
+		}),
+		ctx: ctx,
+	}
+	var err error
+	if len(s.Matchers) == 0 {
+		return nil, trace.NotFound("no matchers present")
+	}
+
+	s.cloudWatcher, err = server.NewCloudWatcher(s.ctx, s.Matchers, s.Clients)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	s.ec2Installer = server.NewSSMInstaller(server.SSMInstallerConfig{
+		Emitter: cfg.Emitter,
+	})
+
+	return s, nil
+}
+
+func (s *Server) filterExistingNodes(instances *server.EC2Instances) {
+	nodes := s.NodeWatcher.GetNodes(func(n services.Node) bool {
+		labels := n.GetAllLabels()
+		_, accountOK := labels[types.AWSAccountIDLabel]
+		_, instanceOK := labels[types.AWSInstanceIDLabel]
+		return accountOK && instanceOK
+	})
+
+	var filtered []*ec2.Instance
+outer:
+	for _, inst := range instances.Instances {
+		for _, node := range nodes {
+			match := types.MatchLabels(node, map[string]string{
+				types.AWSAccountIDLabel:  instances.AccountID,
+				types.AWSInstanceIDLabel: aws.StringValue(inst.InstanceId),
+			})
+			if match {
+				continue outer
+			}
+		}
+		filtered = append(filtered, inst)
+	}
+	instances.Instances = filtered
+}
+
+func (s *Server) handleInstances(instances *server.EC2Instances) error {
+	client, err := s.Clients.GetAWSSSMClient(instances.Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.filterExistingNodes(instances)
+	if len(instances.Instances) == 0 {
+		s.log.Debugf("All fetched nodes already enrolled.")
+		return nil
+	}
+	instIDs := make([]string, len(instances.Instances))
+	for idx, inst := range instances.Instances {
+		instIDs[idx] = aws.StringValue(inst.InstanceId)
+	}
+	s.log.Debugf("Running Teleport installation on these instances: AccountID: %s, Instances: %v", instances.AccountID, instIDs)
+	req := server.SSMRunRequest{
+		DocumentName: instances.DocumentName,
+		SSM:          client,
+		Instances:    instances.Instances,
+		Params:       instances.Parameters,
+		Region:       instances.Region,
+		AccountID:    instances.AccountID,
+	}
+	return trace.Wrap(s.ec2Installer.Run(s.ctx, req))
+}
+
+func (s *Server) handleEC2Discovery() {
+	go s.cloudWatcher.Run()
+	for {
+		select {
+		case instances := <-s.cloudWatcher.InstancesC:
+			instIDs := make([]string, len(instances.Instances))
+			for idx, inst := range instances.Instances {
+				instIDs[idx] = aws.StringValue(inst.InstanceId)
+			}
+			s.log.Debugf("EC2 instances discovered (AccountID: %s, Instances: %v), starting installation", instances.AccountID, instIDs)
+			if err := s.handleInstances(&instances); err != nil {
+				if trace.IsNotFound(err) {
+					s.log.Debug("All discovered EC2 instances are already part of the cluster.")
```

Comment:
It looks like `handleInstances` returns `nil` when all instances are already enrolled, not `NotFound` error.